### PR TITLE
 feat(forms): relation support

### PR DESCRIPTION
**Pull request by zouxuoz**

### CODE REVIEW

File: packages/forms/README.md [hunk 1]
```diff
@@ -335,3 +347,221 @@ render(
...
+  relation: {
+    scalar: 'Relation Scalar Value',
+  },
+  relationList: [{
+    scalar: 'Relation List Scalar Value',
+  }],
+};
+
+render(
+  <SchemaContext.Provider value={SCHEMA}>
```
Comment (ej9x) [inline packages/forms/README.md]: Needs spaces `value={ SCHEMA }`

File: packages/forms/src/types.js [hunk 2]
```diff
@@ -1,4 +1,5 @@
 // @flow
+import * as React from 'react';
```
Comment (ej9x) [inline packages/forms/src/types.js]: Why do we need `import *` here?

File: packages/forms/src/types.js [hunk 3]
```diff
@@ -26,8 +27,12 @@ type TableSchema = {
 
 type Schema = Array<TableSchema>;
 
-type SchemaContextValue = {
-  schema: Schema,
+type SchemaContextValue = Schema;
+
+type RenderableProps = {
+  component?: React.ComponentType<*>,
```
Comment (ej9x) [inline packages/forms/src/types.js]: `React.ComponentType<*>` means that all components in this prop should be the same interfaces. Is this expected behavior?

File: packages/forms/src/types.js [hunk 4]
```diff
@@ -26,8 +27,12 @@ type TableSchema = {
...
+  children?: ((props: Object) => React.Node) | React.Node,
+  render?: (props: Object) => React.Node
```
Comment (ej9x) [inline packages/forms/src/types.js]: Replace `React.Node` to `React$Node`. Then you can escape `import * as React`.

File: packages/forms/src/utils/getFieldSchemaName.js [hunk 5]
```diff
@@ -0,0 +1,13 @@
+// @flow
+const getFieldSchemaName = (name: string): string => {
+  let clearedName = name.replace(/\[\d+\]/g, '');
+
+  if (/\.?([^.]+)$/.test(clearedName)) {
+    // $FlowFixMe
+    clearedName = RegExp.$1;
```
Comment (ej9x) [inline packages/forms/src/utils/getFieldSchemaName.js]: Not clearly behavior. Maybe it's worth adding some unit-tests for this.

File: packages/forms/src/utils/renderComponent.js [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+  }
+
+  if (typeof children === 'function') {
+    rendered = children(rest);
+  } else {
+    rendered = children;
+  }
+
+  return rendered;
+};
```
Comment (ej9x) [inline packages/forms/src/utils/renderComponent.js]: What do you think about adding warnings in cases when user passed `component` prop with `children` or `render` props.

File: packages/forms/src/utils/getFieldSchemaName.js (hunk 5 above)
Comment (zouxuoz) [reply packages/forms/src/utils/getFieldSchemaName.js]: @ej9x Positive cases already tested by high-level tests

File: packages/forms/src/utils/renderComponent.js (hunk 6 above)
Comment (zouxuoz) [reply packages/forms/src/utils/renderComponent.js]: @ej9x good idea

File: packages/forms/src/types.js (hunk 4 above)
Comment (zouxuoz) [reply packages/forms/src/types.js]: @ej9x fixed

File: packages/forms/src/types.js (hunk 3 above)
Comment (zouxuoz) [reply packages/forms/src/types.js]: @ej9x fixed

File: packages/forms/src/types.js (hunk 2 above)
Comment (zouxuoz) [reply packages/forms/src/types.js]: @ej9x fixed

File: packages/forms/README.md (hunk 1 above)
Comment (zouxuoz) [reply packages/forms/README.md]: @ej9x fixed

File: packages/forms/src/utils/renderComponent.js [hunk 7]
```diff
@@ -0,0 +1,31 @@
+// @flow
+import * as React from 'react';
```
[the remark this answers is not in the document]
Comment (zouxuoz) [reply packages/forms/src/utils/renderComponent.js]: fixed

File: packages/forms/src/utils/withFieldSchema.js [hunk 8]
```diff
@@ -2,28 +2,31 @@
 import * as React from 'react';
```
[the remark this answers is not in the document]
Comment (zouxuoz) [reply packages/forms/src/utils/withFieldSchema.js]: fixed

File: packages/forms/src/Field.js [hunk 9]
```diff
@@ -21,7 +21,7 @@ class Field extends React.Component<FieldProps> {
       R.dissoc ('fieldSchema'),
     ),
     R.assoc('validate', createValidate(this.props.fieldSchema)),
-    R.assoc('name', R.defaultTo(this.props.fieldSchema.name)(this.props.name)),
+    R.assoc('name', this.props.name || this.props.name || this.props.fieldSchema.name),
```
Comment (vorobeez) [inline packages/forms/src/Field.js]: Why do you repeat `this.props.name` two times?

**Pull request merged**